### ISSUE TITLE
JGit stops reporting allow-tip-sha1-in-want and/or allow-reachable-sha1-in-want once uploadpack.allowAnySHA1InWant is enabled

### ISSUE THREAD

**Issue by pszlazak** [eclipse-jgit/jgit#68]
### Version

v6.9.0.202403050737-r-23-gc0b415fb0 (bundled with Gerrit 3.10.0)

### Operating System

Linux/Unix

### Bug description

JGit properly advertises following [capabilities](https://git-scm.com/docs/protocol-capabilities) if following options are enabled:
1. [`allow-tip-sha1-in-want`](https://git-scm.com/docs/protocol-capabilities#_allow_tip_sha1_in_want) if [`uploadpack.allowTipSHA1InWant`](https://git-scm.com/docs/git-config/#Documentation/git-config.txt-uploadpackallowTipSHA1InWant) is enabled
1. [`allow-reachable-sha1-in-want`](https://git-scm.com/docs/protocol-capabilities#_allow_reachable_sha1_in_want) if [`uploadpack.allowReachableSHA1InWant`](https://git-scm.com/docs/git-config/#Documentation/git-config.txt-uploadpackallowReachableSHA1InWant) is enabled
1. [`allow-tip-sha1-in-want`](https://git-scm.com/docs/protocol-capabilities#_allow_tip_sha1_in_want) and [`allow-reachable-sha1-in-want`](https://git-scm.com/docs/protocol-capabilities#_allow_reachable_sha1_in_want) if both options are enabled

When [`uploadpack.allowAnySHA1InWant`](https://git-scm.com/docs/git-config/#Documentation/git-config.txt-uploadpackallowAnySHA1InWant) is also enabled, capabilities are no longer advertised (all are gone).

### Actual behavior

If [`uploadpack.allowAnySHA1InWant`](https://git-scm.com/docs/git-config/#Documentation/git-config.txt-uploadpackallowAnySHA1InWant) is enabled, no [`allow-tip-sha1-in-want`](https://git-scm.com/docs/protocol-capabilities#_allow_tip_sha1_in_want) and/or [`allow-reachable-sha1-in-want`](https://git-scm.com/docs/protocol-capabilities#_allow_reachable_sha1_in_want) is advertised by JGit in response for `git-upload-pack` request.

### Expected behavior

JGit continues to report `allow-XXX` capabilities even if [`uploadpack.allowAnySHA1InWant`](https://git-scm.com/docs/git-config/#Documentation/git-config.txt-uploadpackallowAnySHA1InWant) is enabled.

Moreover - as there is not dedicated `allow-any-sha1-in-want` capability, if only [`uploadpack.allowAnySHA1InWant`](https://git-scm.com/docs/git-config/#Documentation/git-config.txt-uploadpackallowAnySHA1InWant) is enabled, other two capabilities should be advertised.

### Relevant log output

Config file with `allowTipSHA1InWant` enabled (only relevant options are shown):
```text
[uploadpack]
    allowTipSHA1InWant = true
    allowReachableSHA1InWant = false
    allowAnySHA1InWant = false
```

Advertised capabilities -> `allow-tip-sha1-in-want` visible:
```shell
$ curl -s http://localhost:8080/Test-project/info/refs?service=git-upload-pack -o - | head -2
001e# service=git-upload-pack
000001082163515068449cf60cb0321bef5adf0e72bdd7db HEAD include-tag multi_ack_detailed multi_ack ofs-delta side-band side-band-64k thin-pack no-progress shallow no-done allow-tip-sha1-in-want agent=JGit/v6.9.0.202403050737-r-23-gc0b415fb0 symref=HEAD:refs/heads/master
                                                                                                                                                                       ^^^^^^^^^^^^^^^^^^^^^^
```

Config file with `allowReachableSHA1InWant` enabled (only relevant options are shown):
```text
[uploadpack]
    allowTipSHA1InWant = false
    allowReachableSHA1InWant = true
    allowAnySHA1InWant = false
```

Advertised capabilities -> `allow-reachable-sha1-in-want` visible:
```shell
$ curl -s http://localhost:8080/Test-project/info/refs?service=git-upload-pack -o - | head -2
001e# service=git-upload-pack
0000010e2163515068449cf60cb0321bef5adf0e72bdd7db HEAD include-tag multi_ack_detailed multi_ack ofs-delta side-band side-band-64k thin-pack no-progress shallow no-done allow-reachable-sha1-in-want agent=JGit/v6.9.0.202403050737-r-23-gc0b415fb0 symref=HEAD:refs/heads/master
                                                                                                                                                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Config file with `allowTipSHA1InWant` and `allowAnySHA1InWant` enabled (only relevant options are shown):
```text
[uploadpack]
    allowTipSHA1InWant = true
    allowReachableSHA1InWant = false
    allowAnySHA1InWant = true
```

Advertised capabilities -> `allow-tip-sha1-in-want` **not** visible:
```shell
$ curl -s http://localhost:8080/Test-project/info/refs?service=git-upload-pack -o - | head -2
001e# service=git-upload-pack
000000f12163515068449cf60cb0321bef5adf0e72bdd7db HEAD include-tag multi_ack_detailed multi_ack ofs-delta side-band side-band-64k thin-pack no-progress shallow no-done agent=JGit/v6.9.0.202403050737-r-23-gc0b415fb0 symref=HEAD:refs/heads/master
```

Config file with `allowReachableSHA1InWant` and `allowAnySHA1InWant` enabled (only relevant options are shown):
```text
[uploadpack]
    allowTipSHA1InWant = false
    allowReachableSHA1InWant = true
    allowAnySHA1InWant = true
```

Advertised capabilities -> `allow-reachable-sha1-in-want` **not** visible:
```shell
$ curl -s http://localhost:8080/Test-project/info/refs?service=git-upload-pack -o - | head -2
001e# service=git-upload-pack
000000f12163515068449cf60cb0321bef5adf0e72bdd7db HEAD include-tag multi_ack_detailed multi_ack ofs-delta side-band side-band-64k thin-pack no-progress shallow no-done agent=JGit/v6.9.0.202403050737-r-23-gc0b415fb0 symref=HEAD:refs/heads/master
```

Config file with `allowTipSHA1InWant` and `allowReachableSHA1InWant` enabled (only relevant options are shown):
```text
[uploadpack]
    allowTipSHA1InWant = true
    allowReachableSHA1InWant = true
    allowAnySHA1InWant = false
```

Advertised capabilities -> `allow-tip-sha1-in-want` and `allow-reachable-sha1-in-want` visible:
```shell
$ curl -s http://localhost:8080/Test-project/info/refs?service=git-upload-pack -o - | head -2
001e# service=git-upload-pack
000001252163515068449cf60cb0321bef5adf0e72bdd7db HEAD include-tag multi_ack_detailed multi_ack ofs-delta side-band side-band-64k thin-pack no-progress shallow no-done allow-tip-sha1-in-want allow-reachable-sha1-in-want agent=JGit/v6.9.0.202403050737-r-23-gc0b415fb0 symref=HEAD:refs/heads/master
                                                                                                                                                                       ^^^^^^^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Config file with `allowTipSHA1InWant`, `allowReachableSHA1InWant` and `allowAnySHA1InWant` enabled (only relevant options are shown):
```text
[uploadpack]
    allowTipSHA1InWant = true
    allowReachableSHA1InWant = true
    allowAnySHA1InWant = true
```

Advertised capabilities -> `allow-tip-sha1-in-want` and `allow-reachable-sha1-in-want` **not** visible:
```shell
$ curl -s http://localhost:8080/Test-project/info/refs?service=git-upload-pack -o - | head -2
001e# service=git-upload-pack
000000f12163515068449cf60cb0321bef5adf0e72bdd7db HEAD include-tag multi_ack_detailed multi_ack ofs-delta side-band side-band-64k thin-pack no-progress shallow no-done agent=JGit/v6.9.0.202403050737-r-23-gc0b415fb0 symref=HEAD:refs/heads/master
```

Config file with only `allowAnySHA1InWant` enabled (only relevant options are shown):
```text
[uploadpack]
    allowTipSHA1InWant = false
    allowReachableSHA1InWant = false
    allowAnySHA1InWant = true
```
Advertised capabilities -> `allow-tip-sha1-in-want` and `allow-reachable-sha1-in-want` **not** visible:
```shell
$ curl -s http://localhost:8080/Test-project/info/refs?service=git-upload-pack -o - | head -2
001e# service=git-upload-pack
000000f12163515068449cf60cb0321bef5adf0e72bdd7db HEAD include-tag multi_ack_detailed multi_ack ofs-delta side-band side-band-64k thin-pack no-progress shallow no-done agent=JGit/v6.9.0.202403050737-r-23-gc0b415fb0 symref=HEAD:refs/heads/master
```

### Other information

Tested with [Gerrit 3.10.0](https://www.gerritcodereview.com/3.10.html)

**Comment by lucamilanesio** [comment]
I believe this is due to how JGit stores the allow flags, compared to CGit.

CGit:
```c
/* Enum for allowed unadvertised object request (UOR) */
enum allow_uor {
        /* Allow specifying sha1 if it is a ref tip. */
        ALLOW_TIP_SHA1 = 0x01,
        /* Allow request of a sha1 if it is reachable from a ref (possibly hidden ref). */
        ALLOW_REACHABLE_SHA1 = 0x02,
        /* Allow request of any sha1. Implies ALLOW_TIP_SHA1 and ALLOW_REACHABLE_SHA1. */
        ALLOW_ANY_SHA1 = 0x07
};
```

JGit:
```java
        /** Policy the server uses to validate client requests */
        public enum RequestPolicy {
                /** Client may only ask for objects the server advertised a reference for. */
                ADVERTISED,

                /**
                 * Client may ask for any commit reachable from a reference advertised by
                 * the server.
                 */
                REACHABLE_COMMIT,

                /**
                 * Client may ask for objects that are the tip of any reference, even if not
                 * advertised.
                 * <p>
                 * This may happen, for example, when a custom {@link RefFilter} is set.
                 *
                 * @since 3.1
                 */
                TIP,

                /**
                 * Client may ask for any commit reachable from any reference, even if that
                 * reference wasn't advertised.
                 *
                 * @since 3.1
                 */
                REACHABLE_COMMIT_TIP,

                /** Client may ask for any SHA-1 in the repository. */
                ANY;
        }
```

With CGit, it is evident that ANY (binary flag = 111) includes TIP (binary flag = 100) and REACHABLE (binary flag = 010), but with JGit it isn't because the enums do not have values.

**Comment by lucamilanesio** [comment]
I've proposed [this fix that aligns the CGit and JGit definitions](https://review.gerrithub.io/c/eclipse-jgit/jgit/+/1202276) and IMHO is low-risk as it does not change anything in the logic of how the enums are used, apart from the more readable expression `RequestPolicy.implies()` in the logic for advertising capabilities.

**Comment by lucamilanesio** [comment]
@pszlazak I have noticed that this bug impacted only Git protocol V0 in JGit: have you tried enabling Git protocol V2?

**Comment by pszlazak** [comment]
@lucamilanesio Correct, `allow-tip-sha1-in-want` and `allow-reachable-sha1-in-want` are capabilities related to V0/V1 Git protocol:  
> this document describes capabilities for versions 0 and 1 of the pack protocol. For version 2, please refer to the [protocol-v2](https://git-scm.com/docs/protocol-v2) doc.

[[source](https://git-scm.com/docs/protocol-capabilities)]

For V2 there is another set of capabilities: https://git-scm.com/docs/gitprotocol-v2#_capabilities  
`uploadpack.allowTipSHA1InWant`, `uploadpack.allowReachableSHA1InWant` and `uploadpack.allowAnySHA1InWant` options do not have impact on them AFAIK.

**Comment by pszlazak** [comment]
@lucamilanesio I managed to test your correction. I cherry-picked it to JGit 6.10 branch, compiled and integrated with Gerrit 3.10.x - issue is fixed. You can now have `uploadpack.allowAnySHA1InWant` set to `true` added to `jgit.config` and `allow-tip-sha1-in-want` and `allow-reachable-sha1-in-want` still are advertised.

Yet in a meantime I had a conversation with CGit maintainers: https://lore.kernel.org/git/pull.1814.git.git.1729112794671.gitgitgadget@gmail.com/T/#u

In a nutshell - in CGit `uploadpack.allowAnySHA1InWant` overrides other two `uploadpack.allowXXXInWant` settings. If set to `false`, it will also disable `uploadpack.allowTipSHA1InWant` and `uploadpack.allowReachableSHA1InWant`. Same goes for `true`.  
But in CGit also order of settings is important - if first you have `uploadpack.allowAnySHA1InWant` set to `false`, still later you can  individually enable `uploadpack.allowTipSHA1InWant` and `uploadpack.allowReachableSHA1InWant`.

I believe it doesn't work the same in JGit. Yet I don't think it's needed, especially that it is only valid for V0/V1 Git protocol.

If you share this opinion and no further changes in JGit are needed, I prepared changes for stable-6.10 branch:
- https://review.gerrithub.io/c/eclipse-jgit/jgit/+/1204171
- https://review.gerrithub.io/c/eclipse-jgit/jgit/+/1204172

FYI - because I lack forge-author permission, I had to reset the author.

**Comment by pszlazak** [comment]
Issue fixed in 1519c147948eb1108bdf45f2aeed84746dacff9c (Gerrit change [eclipse-jgit/jgit/+/1202276](https://review.gerrithub.io/c/eclipse-jgit/jgit/+/1202276)) + tests committed in 5583f6a10eafc8c2627e0fb4833cb8ffe422f69a (Gerrit change [eclipse-jgit/jgit/+/1202480](https://review.gerrithub.io/c/eclipse-jgit/jgit/+/1202480)).

Also cherry-picked to [stable-6.10](https://github.com/eclipse-jgit/jgit/tree/stable-6.10) branch as 5b1513a28d337e7e3453e557ee9dde292678eb81 (Gerrit change [eclipse-jgit/jgit/+/1204171](https://review.gerrithub.io/c/eclipse-jgit/jgit/+/1204171)) + tests committed in 079dbe8ed9e47e44986ce43513f3aad8fa64832a (Gerrit change [eclipse-jgit/jgit/+/1204172](https://review.gerrithub.io/c/eclipse-jgit/jgit/+/1204172)).
